zero/src/main/java/mutiny/zero/internal/IterablePublisher.java

```diff
@@ -0,0 +1,27 @@
+package mutiny.zero.internal;
+
+import static java.util.Objects.requireNonNull;
+
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+
+public class IterablePublisher<T> implements Publisher<T> {
+
+    private final Iterable<T> iterable;
+
+    public IterablePublisher(Iterable<T> iterable) {
+        this.iterable = iterable;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super T> subscriber) {
+        requireNonNull(subscriber, "The subscriber cannot be null");
+        if (!iterable.iterator().hasNext()) {
```

Comment:
Is this optimisation worth it? I've seen a few `Iterables` whose `Iterable.iterator()` call is a bit expensive, so doing it twice could be costly.

Reply:
If the stream has already completed then you should signal it at subscription time.

Reply:
But let's assume this `iterable.iterator()` call is really heavy, so 1) you're blocking on the subscription time, and 2 it may or may not complete depending on a variety of things that may even depend on how many items a subscriber requests. If the subscriber doesn't request any items, then why even start the iterator?

This optimisation may work for JDK collections, which are already materialised, but what if `iterable` was a SQL query? Do you really want to call `iterator()` right here, and twice so?

Reply:
Iterables should be to iterate in-memory data structures.

If you do something that blocks, well... don't use an iterable, use the `Tube` interface.

Reply:
And note that it's not an optimisation per-se.

Reply:
Absolutely, this will be specified in the API.